pareto/tests/test_utilities.py

```diff
@@ -0,0 +1,337 @@
+#####################################################################################################
+# PARETO was produced under the DOE Produced Water Application for Beneficial Reuse Environmental
+# Impact and Treatment Optimization (PARETO), and is copyright (c) 2021-2023 by the software owners:
+# The Regents of the University of California, through Lawrence Berkeley National Laboratory, et al.
+# All rights reserved.
+#
+# NOTICE. This Software was developed under funding from the U.S. Department of Energy and the U.S.
+# Government consequently retains certain rights. As such, the U.S. Government has been granted for
+# itself and others acting on its behalf a paid-up, nonexclusive, irrevocable, worldwide license in
+# the Software to reproduce, distribute copies to the public, prepare derivative works, and perform
+# publicly and display publicly, and to permit others to do so.
+#####################################################################################################
+"""
+Test utilities
+"""
+
+from pareto.strategic_water_management.strategic_produced_water_optimization import (
+    WaterQuality,
+    create_model,
+    Objectives,
+    solve_model,
+    PipelineCost,
+    PipelineCapacity,
+    RemovalEfficiencyMethod,
+)
+from pareto.utilities.get_data import get_data
+from pareto.utilities.results import (
+    generate_report,
+    PrintValues,
+    OutputUnits,
+    is_feasible,
+    nostdout,
+)
+from importlib import resources
+
+from pyomo.environ import value
+
+# Modules to test:
+from pareto.utilities.bounding_functions import VariableBounds
+from pareto.utilities.model_modifications import free_variables
+from pareto.utilities.model_modifications import deactivate_slacks
+
+# import pytest
+import pytest
+
+
+############################
+def build_strategic_model(config_dict):
```

Comment:
Two things:

1) Is there a reason you aren't using `@pytest.fixture(scope="module")` like we do in `test_strategic_model.py`?
2) This function doesn't just build the strategic model, it also solves it. So I recommend changing the name of the function to reflect that fact.

Reply:
1. I don't know what that first part is for. It can't be important though, or someone would have mentioned it earlier.
2. I need the strategic model with its quality block. If I could get it *without* solving the model, I would. The naming serves its purpose, in that respect.

Reply:
@Tominapa your reasoning on 1 is fallacious. I was the second person to review this PR - you can't dismiss my question as unimportant simply because the first reviewer didn't mention it in their review. That said, while you could use a pytest fixture here, I don't think you need to - I was just wondering if there was a reason why you had chosen not to. You can learn more about fixtures [here](https://docs.pytest.org/en/6.2.x/fixture.html) and you can look at some of our other unit tests for examples of when/how to use them.

On 2 - as it is currently written, the function `build_strategic_model` allows you to build and solve the strategic model with whatever option you want for quality (as specified in the `config_dict` argument that the function takes). I understand why you need to solve the model in this function, so I suggest changing its name to `build_and_solve_strategic_model` to more accurately describe its behavior. In other parts of our codebase, functions that say "build" in the name generally just build a model without solving. So since this function both builds and also solves the model, you can prevent some potential confusion for other developers by changing the name to reflect that. This is admittedly a somewhat minor thing, but it's a change that will help anyone who reads this code somewhere down the line.

Reply:
1. Thanks for the pointer to the docs. I figured out why you are asking for the change based on the context there. It should be the same with the change, benefiting from any marginal test time reductions.
2. I see your point, but the key from my point of view is that I do not _want_ to solve the model at this stage. If I could construct the quality model without a solve call, I would. This would require a change within PARETO itself, which I would rather avoid - at least for the moment. I would also suggest that this function contains neither build nor solve calls directly. A more accurate name, by this logic, would simply be "fetch_strategic_model()"

Allow me some time to complete these changes. I have other priorities competing today.

Reply:
@Tominapa:

1) to be clear, I don't think you _need_ to restructure this code to use a fixture. Feel free to do so if you wish, but I don't think it's strictly necessary.

2) Your function `build_strategic_model()` calls `create_model()` and then `solve_model()` - these are the functions that build the strategic model and then solve it, respectively, so I'm not sure how you can argue that `build_strategic_model()` "contains neither build nor solve calls directly". I understand that the fact that you need to call `solve()` to get the quality model is an artifact of how the strategic model is designed. I'm just saying that it's easier to read code where function names describe as closely as possible the actions of the function (even if some of those actions are side effects). To be clear, this is a relatively minor thing, and I'm not going to block you from merging this PR or anything if you don't want to change the name of the function, I just want to explain my opinion clearly.

Reply:
1. Done.
2. Indeed, the build and solve calls are in those other functions. I don't want to give the impression that I am recreating those capabilities here. I'm just calling an existing function to get a model on which to run tests. The name is now "fetch_strategic_model()" to be compatible with your guidelines.